### PR TITLE
Add VPConstraints for Floating Point nodes

### PR DESCRIPTION
**Pull request by shivammittal99**
This PR is a part of my [GSOC'18 Project](https://summerofcode.withgoogle.com/projects/5175053981843456).

1. Added some simplifications for FP nodes : fadd, fsub, fmul, fneg, dadd, dsub, dmul, dneg. ( #2599 )
2. Added Value Propagation constraints : VPFloatConstraint, VPFloatConst, VPFloatRange, VPDoubleConstriant, VPDoubleConst, VPDoubleRange

### CODE REVIEW

**Comment by 0xdaryl** [comment]
@shivammittal99 , if this is ready for a review from @vijaysun-omr please remove the WIP: prefix.

**Comment by shivammittal99** [comment]
Sorry about that. Done

**Comment by vijaysun-omr** [comment]
I would also like a review from @jdmpapin

**Comment by vijaysun-omr** [comment]
How has this code been tested ?

**Comment by vijaysun-omr** [comment]
@genie-omr build all

File: compiler/optimizer/VPConstraint.hpp [hunk 1]
```diff
@@ -310,6 +331,34 @@ class VPConstraint
       return diff;
       }
 
+   float floatAddWithOverflow(float a, float b, bool& overflow)
+      {
+      float sum = TR::Compiler->arith.floatAddFloat(a, b);
+      overflow = (sum == FLOAT_POS_INFINITY) || (sum == FLOAT_NEG_INFINITY);
```
Comment (vijaysun-omr) [inline compiler/optimizer/VPConstraint.hpp]: Could you please elaborate on your reasoning behind considering this an adequate condition for detecting overflow in these new routines you added in this commit ?

File: compiler/optimizer/VPHandlers.cpp [hunk 2]
```diff
@@ -5966,7 +6082,19 @@ TR::Node *constrainAdd(OMR::ValuePropagation *vp, TR::Node *node)
...
+      if (rhs->asFloatConst())
+         {
+         if (rhs->asFloatConst()->getFloat() > FLOAT_NEG_INFINITY &&
+             rhs->asFloatConst()->getFloat() < FLOAT_POS_INFINITY)
+            constraint = TR::VPEqual::create(vp, rhs->asFloatConst()->getFloat());
+         }
+      else if (rhs->asDoubleConst())
+         {
+         if (rhs->asDoubleConst()->getDouble() > FLOAT_NEG_INFINITY &&
+             rhs->asDoubleConst()->getDouble() < FLOAT_POS_INFINITY)
```
Comment (vijaysun-omr) [inline compiler/optimizer/VPHandlers.cpp]: These 2 lines look like typos. Should you be using the DOUBLE* versions of infinity ?

File: compiler/optimizer/VPConstraint.hpp (hunk 1 above)
Comment (shivammittal99) [reply compiler/optimizer/VPConstraint.hpp]: According to IEEE 754, whenever there is an overflow the result returned is INFINITY

**Comment by charliegracie** [comment]
ping @shivammittal99

**Comment by mstoodle** [comment]
Hi @shivammittal99 can you estimate when you might be able to fix the problem found in CI testing?

**Comment by shivammittal99** [comment]
@mstoodle I am not sure why this is erroring for both float and double. `nextafter(double, double)` is also supposed to be part of C99 and `math.h` is getting included. There is a small change for float version i.e `nextafterf(float, float)`, but it should not give error for double version.

**Comment by vijaysun-omr** [comment]
@shivammittal99 any update ?

**Comment by vijaysun-omr** [comment]
@0xdaryl any suggestions on how to proceed with this PR ?

**Comment by 0xdaryl** [comment]
I am closing this PR now given its age and the lack of responses for updates from committers.  Thanks for the contribution.

**Pull request closed without merging**